File: packages/polygon-packer/src/wasm-packer.ts
import { get_u16_from_u32, abs_polygon_area } from 'wasm-nesting';
import GeneticAlgorithm from './genetic-algorithm';
import NFPStore from './nfp-store';
import { f32, NestConfig, SourceItem } from './types';
import { readUint32FromF32, generateTree, generateBounds, deserializeConfig } from './helpers';
import { BoundRectF32 } from './geometry';
import PolygonNode from './polygon-node';

export default class WasmPacker {

    #binNode: PolygonNode = null;

    #binArea: f32 = 0;

    #binBounds: BoundRectF32 = null;

    #resultBounds: BoundRectF32 = null;

    #best: Float32Array = null;

    #nodes: PolygonNode[] = [];

    #config: NestConfig = null;
    // progressCallback is called when progress is made
    // displayCallback is called when a new placement has been made
    public init(
        configuration: number,
        polygonData: Float32Array,
        sizes: Uint16Array,
    ): void {
        let offset = 0;
        const polygons: Float32Array[] = [];

        sizes.forEach((size: number) => {
            polygons.push(polygonData.subarray(offset, offset + size));
            offset += size;
        });

        const binPolygon = polygons.pop();
        this.#config = deserializeConfig(configuration);
        const binData = generateBounds(binPolygon, this.#config.spacing, this.#config.curveTolerance);

        this.#binNode = binData.binNode;
        this.#binBounds = binData.bounds;
        this.#resultBounds = binData.resultBounds;
        this.#binArea = binData.area;
        this.#nodes = generateTree(polygons, this.#config.spacing, this.#config.curveTolerance);

        GeneticAlgorithm.instance.init(this.#nodes, this.#resultBounds, this.#config);
    }

    public getPairs(): Float32Array {
        const individual = GeneticAlgorithm.instance.getIndividual(this.#nodes);
        NFPStore.instance.init(this.#nodes, this.#binNode, this.#config, individual.source, individual.placement, individual.rotation);

        const pairs = NFPStore.instance.nfpPairs;

        // Serialize pairs: count (f32) + [size (f32) + data] for each pair
        let totalSize = 1; // count as f32
        for (const pair of pairs) {
            totalSize += 1 + pair.length; // size as f32 + data
        }

        const buffer = new Float32Array(totalSize);
        let offset = 0;

        // Write count (as bits of u32)
        buffer[offset] = new Float32Array(new Uint32Array([pairs.length]).buffer)[0];
        offset += 1;

        // Write each pair
        for (const pair of pairs) {
            buffer[offset] = new Float32Array(new Uint32Array([pair.length]).buffer)[0];
            offset += 1;

            buffer.set(pair, offset);
            offset += pair.length;
        }

        return buffer;
    }

    public getPlacementData(generatedNfp: ArrayBuffer[]): Uint8Array {
        NFPStore.instance.update(generatedNfp);

        return NFPStore.instance.getPlacementData(this.#nodes, this.#binArea);
    }

    public getPlacemehntResult(placements: ArrayBuffer[]): Uint8Array {
        let placementsData: Float32Array = new Float32Array(placements[0]);
        let currentPlacement: Float32Array = null;

        GeneticAlgorithm.instance.updateFitness(NFPStore.instance.phenotypeSource, placementsData[0]);

        for (let i = 1; i < placements.length; ++i) {
            currentPlacement = new Float32Array(placements[i]);
            if (currentPlacement[0] < placementsData[0]) {
                placementsData = currentPlacement;
            }
        }

        let result = null;
        let numParts: number = 0;
        let numPlacedParts: number = 0;
        let placePerecntage: number = 0;

        if (!this.#best || placementsData[0] < this.#best[0]) {
            this.#best = placementsData;

            const binArea: number = Math.abs(this.#binArea);
            const placementCount = placementsData[1];
            let placedCount: number = 0;
            let placedArea: number = 0;
            let totalArea: number = 0;
            let pathId: number = 0;
            let itemData: number = 0;
            let offset: number = 0;
            let size: number = 0;

            for (let i = 0; i < placementCount; ++i) {
                totalArea += binArea;
                itemData = readUint32FromF32(placementsData, 2 + i);
                offset = get_u16_from_u32(itemData, 1);
                size = get_u16_from_u32(itemData, 0);
                placedCount += size;

                for (let j = 0; j < size; ++j) {
                    pathId = get_u16_from_u32(readUint32FromF32(placementsData, offset + j), 1);
                    placedArea += abs_polygon_area(this.#nodes[pathId].memSeg);
                }
            }

            numParts = NFPStore.instance.placementCount;
            numPlacedParts = placedCount;
            placePerecntage = placedArea / totalArea;
            result = { placementsData };
        }

        const serializedResult = WasmPacker.serializePlacementResult(
            placePerecntage,
            numPlacedParts,
            numParts,
            this.#config.rotations,
            result !== null,
            this.#binBounds,
            result ? WasmPacker.convertPolygonNodesToSourceItems(this.#nodes) : [],
            result ? result.placementsData : new Float32Array(0)
        );


        return serializedResult;
    }

    public stop(): void {
        this.#nodes = [];
        this.#best = null;
        this.#binNode = null;
        GeneticAlgorithm.instance.clean();
        NFPStore.instance.clean();
    }

    public get pairCount(): number {
        return NFPStore.instance.nfpPairsCount;
    }

    static serializePlacementResult(
        placePercentage: number,
        numPlacedParts: number,
        numParts: number,
        angleSplit: number,
        hasResult: boolean,
        bounds: BoundRectF32,
        sources: SourceItem[],
        placementsData: Float32Array
    ): Uint8Array {
        // Serialize sources to get the size
        const serializedSources = WasmPacker.serializeSourceItems(sources);
        const sourcesSize = serializedSources.byteLength;
        const placementsDataSize = placementsData.byteLength;

        // Calculate total buffer size:
        // placePercentage (4) + numPlacedParts (2) + numParts (2) + angleSplit (1) + hasResult (1)
        // + boundsX (4) + boundsY (4) + boundsWidth (4) + boundsHeight (4)
        // + sourcesSize (4) + placementsDataSize (4)
        // + serializedSources + serializedPlacementsData
        const headerSize = 4 + 2 + 2 + 1 + 1 + 4 + 4 + 4 + 4 + 4 + 4;
        const totalSize = headerSize + sourcesSize + placementsDataSize;

        const buffer = new ArrayBuffer(totalSize);
        const view = new DataView(buffer);
        let offset = 0;

        // Write header
        view.setFloat32(offset, placePercentage, true);
        offset += 4;

        view.setUint16(offset, numPlacedParts, true);
        offset += 2;

        view.setUint16(offset, numParts, true);
        offset += 2;

        view.setUint8(offset, angleSplit);
        offset += 1;

        view.setUint8(offset, hasResult ? 1 : 0);
        offset += 1;

        view.setFloat32(offset, bounds.x, true);
        offset += 4;

        view.setFloat32(offset, bounds.y, true);
        offset += 4;

        view.setFloat32(offset, bounds.width, true);
        offset += 4;

        view.setFloat32(offset, bounds.height, true);
        offset += 4;

        view.setUint32(offset, sourcesSize, true);
        offset += 4;

        view.setUint32(offset, placementsDataSize, true);
        offset += 4;

        // Write serialized sources
        new Uint8Array(buffer, offset, sourcesSize).set(serializedSources);
        offset += sourcesSize;

        // Write placements data
        new Uint8Array(buffer, offset, placementsDataSize).set(new Uint8Array(placementsData.buffer));

        return new Uint8Array(buffer);
    }

    private static convertPolygonNodesToSourceItems(nodes: PolygonNode[]): SourceItem[] {
        return nodes.map(node => WasmPacker.convertPolygonNodeToSourceItem(node));
    }

    private static convertPolygonNodeToSourceItem(node: PolygonNode): SourceItem {
        return {
            source: node.source,
            children: node.children.map(child => WasmPacker.convertPolygonNodeToSourceItem(child))
        };
    }

    private static calculateSourceItemsSize(items: SourceItem[]): number {
        return items.reduce((total, item) => {
            // Each item: u16 (source) + u16 (children count) = 4 bytes
            const itemSize = Uint16Array.BYTES_PER_ELEMENT * 2;
            return total + itemSize + WasmPacker.calculateSourceItemsSize(item.children);
        }, 0);
    }

    private static serializeSourceItemsInternal(items: SourceItem[], view: DataView, offset: number): number {
        let currentOffset = offset;

        for (const item of items) {
            // Write source (u16)
            view.setUint16(currentOffset, item.source, true);
            currentOffset += Uint16Array.BYTES_PER_ELEMENT;

            // Write children count (u16)
            view.setUint16(currentOffset, item.children.length, true);
            currentOffset += Uint16Array.BYTES_PER_ELEMENT;

            // Recursively serialize children
            currentOffset = WasmPacker.serializeSourceItemsInternal(item.children, view, currentOffset);
        }

        return currentOffset;
    }

    private static serializeSourceItems(items: SourceItem[]): Uint8Array {
        // Calculate total size: u16 (count) + items data
        const itemsSize = WasmPacker.calculateSourceItemsSize(items);
        const totalSize = Uint16Array.BYTES_PER_ELEMENT + itemsSize;

        const buffer = new ArrayBuffer(totalSize);
        const view = new DataView(buffer);

        // Write items count
        view.setUint16(0, items.length, true);

        // Serialize items
        WasmPacker.serializeSourceItemsInternal(items, view, Uint16Array.BYTES_PER_ELEMENT);

        return new Uint8Array(buffer);
    }
}

File: packages/wasm-nesting/src/wasm_packer.rs
use crate::{
    genetic_algorithm::GeneticAlgorithm,
    geometry::bound_rect::BoundRect,
    nest_config::NestConfig,
    nesting::{nfp_store::NFPStore, polygon_node::PolygonNode},
    utils::{bit_ops::get_u16, number::Number},
};

pub struct WasmPacker {
    bin_node: Option<PolygonNode>,
    bin_area: f32,
    bin_bounds: Option<BoundRect<f32>>,
    result_bounds: Option<BoundRect<f32>>,
    best: Option<Vec<f32>>,
    nfp_store: NFPStore,
    nodes: Vec<PolygonNode>,
    config: NestConfig,
}

impl WasmPacker {
    pub fn new() -> Self {
        WasmPacker {
            bin_node: None,
            bin_area: 0.0,
            bin_bounds: None,
            result_bounds: None,
            best: None,
            nfp_store: NFPStore::new(),
            nodes: Vec::new(),
            config: NestConfig::new(),
        }
    }

    pub fn init(&mut self, configuration: u32, polygon_data: &[f32], sizes: &[u16]) {
        let mut offset = 0;
        let mut polygons: Vec<&[f32]> = Vec::new();

        for &size in sizes {
            let size_usize = size as usize;
            polygons.push(&polygon_data[offset..offset + size_usize]);
            offset += size_usize;
        }

        // Last polygon is the bin
        let bin_polygon = polygons.pop().unwrap();

        // Deserialize config
        self.config.deserialize(configuration);

        // Generate bounds for bin
        let bin_data = Self::generate_bounds_internal(
            bin_polygon,
            self.config.spacing as i32,
            self.config.curve_tolerance,
        );

        self.bin_node = Some(bin_data.bin_node);
        self.bin_bounds = Some(bin_data.bounds);
        self.result_bounds = Some(bin_data.result_bounds);
        self.bin_area = bin_data.area;

        // Generate tree for other polygons
        self.nodes = Self::generate_tree_internal(
            &polygons,
            self.config.spacing as i32,
            self.config.curve_tolerance,
        );

        // Initialize genetic algorithm
        GeneticAlgorithm::with_instance(|ga| {
            ga.init(
                &self.nodes,
                self.result_bounds.as_ref().unwrap(),
                &self.config,
            );
        });
    }

    pub fn get_pairs(&mut self) -> Vec<u8> {
        let individual = GeneticAlgorithm::with_instance(|ga| ga.get_individual(&self.nodes))
            .expect("Failed to get individual");

        self.nfp_store.init(
            &self.nodes,
            self.bin_node.as_ref().unwrap(),
            &self.config,
            individual.source(),
            individual.placement(),
            individual.rotation(),
        );

        let pairs = self.nfp_store.nfp_pairs();

        // Serialize pairs: count (u32) + [size (u32) + data] for each pair
        let mut total_size = 4; // count
        for pair in pairs {
            total_size += 4 + pair.len(); // size + data
        }

        let mut buffer = vec![0u8; total_size];
        let mut offset = 0;

        // Write count
        buffer[offset..offset + 4].copy_from_slice(&(pairs.len() as u32).to_le_bytes());
        offset += 4;

        // Write each pair
        for pair in pairs {
            buffer[offset..offset + 4].copy_from_slice(&(pair.len() as u32).to_le_bytes());
            offset += 4;

            buffer[offset..offset + pair.len()].copy_from_slice(pair);
            offset += pair.len();
        }

        buffer
    }

    pub fn get_placement_data(&mut self, generated_nfp: Vec<Vec<u8>>) -> Vec<u8> {
        self.nfp_store.update(generated_nfp);
        self.nfp_store
            .get_placement_data(&self.nodes, self.bin_area)
    }

    pub fn get_placement_result(&mut self, placements: Vec<Vec<u8>>) -> Vec<u8> {
        if placements.is_empty() {
            return Vec::new();
        }

        // Convert first placement to f32 slice
        let mut placements_data = Self::bytes_to_f32_vec(&placements[0]);

        GeneticAlgorithm::with_instance(|ga| {
            ga.update_fitness(self.nfp_store.phenotype_source(), placements_data[0]);
        });

        // Find best placement
        for i in 1..placements.len() {
            let current_placement = Self::bytes_to_f32_vec(&placements[i]);
            if current_placement[0] < placements_data[0] {
                placements_data = current_placement;
            }
        }

        let mut num_parts: u16 = 0;
        let mut num_placed_parts: u16 = 0;
        let mut place_percentage: f32 = 0.0;
        let mut has_result = false;

        if self.best.is_none() || placements_data[0] < self.best.as_ref().unwrap()[0] {
            self.best = Some(placements_data.clone());

            let bin_area = self.bin_area.abs();
            let placement_count = placements_data[1] as usize;
            let mut placed_count: u16 = 0;
            let mut placed_area: f32 = 0.0;
            let mut total_area: f32 = 0.0;

            for i in 0..placement_count {
                total_area += bin_area;
                let item_data = Self::read_uint32_from_f32(&placements_data, 2 + i);
                let offset = get_u16(item_data, 1) as usize;
                let size = get_u16(item_data, 0);
                placed_count += size;

                for j in 0..size as usize {
                    let path_data = Self::read_uint32_from_f32(&placements_data, offset + j);
                    let path_id = get_u16(path_data, 1) as usize;
                    placed_area += f32::abs_polygon_area(&self.nodes[path_id].mem_seg) as f32;
                }
            }

            num_parts = self.nfp_store.placement_count() as u16;
            num_placed_parts = placed_count;
            place_percentage = placed_area / total_area;
            has_result = true;
        }

        // Serialize result
        Self::serialize_placement_result(
            place_percentage,
            num_placed_parts,
            num_parts,
            self.config.rotations,
            has_result,
            self.bin_bounds.as_ref().unwrap(),
            if has_result {
                Self::convert_nodes_to_source_items(&self.nodes)
            } else {
                Vec::new()
            },
            if has_result { &placements_data } else { &[] },
        )
    }

    pub fn stop(&mut self) {
        self.nodes.clear();
        self.best = None;
        self.bin_node = None;
        GeneticAlgorithm::with_instance(|ga| {
            ga.clean();
        });
        self.nfp_store.clean();
    }

    pub fn pair_count(&self) -> usize {
        self.nfp_store.nfp_pairs().len()
    }

    // Helper functions

    fn bytes_to_f32_vec(bytes: &[u8]) -> Vec<f32> {
        let len = bytes.len() / 4;
        let mut result = vec![0f32; len];
        unsafe {
            std::ptr::copy_nonoverlapping(
                bytes.as_ptr(),
                result.as_mut_ptr() as *mut u8,
                bytes.len(),
            );
        }
        result
    }

    fn read_uint32_from_f32(data: &[f32], index: usize) -> u32 {
        data[index].to_bits()
    }

    fn generate_bounds_internal(mem_seg: &[f32], spacing: i32, curve_tolerance: f32) -> BoundsData {
        use crate::clipper_wrapper;

        let result = clipper_wrapper::generate_bounds(mem_seg, spacing, curve_tolerance as f64);

        match result {
            Some((bounds, result_bounds, area, node)) => BoundsData {
                bounds,
                result_bounds,
                area: area as f32,
                bin_node: node,
            },
            None => panic!("Failed to generate bounds"),
        }
    }

    fn generate_tree_internal(
        polygons: &[&[f32]],
        spacing: i32,
        curve_tolerance: f32,
    ) -> Vec<PolygonNode> {
        use crate::clipper_wrapper;

        // Flatten polygons into single array with sizes
        let mut total_length = 0;
        let mut sizes = Vec::new();

        for polygon in polygons {
            sizes.push((polygon.len() >> 1) as u16); // Point count
            total_length += polygon.len();
        }

        let mut values = vec![0f32; total_length];
        let mut offset = 0;

        for polygon in polygons {
            values[offset..offset + polygon.len()].copy_from_slice(polygon);
            offset += polygon.len();
        }

        clipper_wrapper::generate_tree(&values, &sizes, spacing, curve_tolerance as f64)
    }

    fn convert_nodes_to_source_items(nodes: &[PolygonNode]) -> Vec<SourceItem> {
        nodes
            .iter()
            .map(|node| Self::convert_node_to_source_item(node))
            .collect()
    }

    fn convert_node_to_source_item(node: &PolygonNode) -> SourceItem {
        SourceItem {
            source: node.source as u16,
            children: Self::convert_nodes_to_source_items(&node.children),
        }
    }

    fn calculate_source_items_size(items: &[SourceItem]) -> usize {
        items.iter().fold(0, |total, item| {
            // Each item: u16 (source) + u16 (children count) = 4 bytes
            let item_size = 4;
            total + item_size + Self::calculate_source_items_size(&item.children)
        })
    }

    fn serialize_source_items_internal(
        items: &[SourceItem],
        buffer: &mut [u8],
        offset: usize,
    ) -> usize {
        let mut current_offset = offset;

        for item in items {
            // Write source (u16)
            buffer[current_offset..current_offset + 2].copy_from_slice(&item.source.to_le_bytes());
            current_offset += 2;

            // Write children count (u16)
            buffer[current_offset..current_offset + 2]
                .copy_from_slice(&(item.children.len() as u16).to_le_bytes());
            current_offset += 2;

            // Recursively serialize children
            current_offset =
                Self::serialize_source_items_internal(&item.children, buffer, current_offset);
        }

        current_offset
    }

    fn serialize_source_items(items: &[SourceItem]) -> Vec<u8> {
        let items_size = Self::calculate_source_items_size(items);
        let total_size = 2 + items_size; // u16 count + items data

        let mut buffer = vec![0u8; total_size];

        // Write items count
        buffer[0..2].copy_from_slice(&(items.len() as u16).to_le_bytes());

        // Serialize items
        Self::serialize_source_items_internal(items, &mut buffer, 2);

        buffer
    }

    fn serialize_placement_result(
        place_percentage: f32,
        num_placed_parts: u16,
        num_parts: u16,
        angle_split: u8,
        has_result: bool,
        bounds: &BoundRect<f32>,
        sources: Vec<SourceItem>,
        placements_data: &[f32],
    ) -> Vec<u8> {
        let serialized_sources = Self::serialize_source_items(&sources);
        let sources_size = serialized_sources.len();
        let placements_data_size = placements_data.len() * 4;

        // Header: 34 bytes
        // placePercentage (4) + numPlacedParts (2) + numParts (2) + angleSplit (1) + hasResult (1)
        // + boundsX (4) + boundsY (4) + boundsWidth (4) + boundsHeight (4)
        // + sourcesSize (4) + placementsDataSize (4)
        let header_size = 34;
        let total_size = header_size + sources_size + placements_data_size;

        let mut buffer = vec![0u8; total_size];
        let mut offset = 0;

        // Write header
        buffer[offset..offset + 4].copy_from_slice(&place_percentage.to_le_bytes());
        offset += 4;

        buffer[offset..offset + 2].copy_from_slice(&num_placed_parts.to_le_bytes());
        offset += 2;

        buffer[offset..offset + 2].copy_from_slice(&num_parts.to_le_bytes());
        offset += 2;

        buffer[offset] = angle_split;
        offset += 1;

        buffer[offset] = if has_result { 1 } else { 0 };
        offset += 1;

        unsafe {
            buffer[offset..offset + 4].copy_from_slice(&bounds.x().to_le_bytes());
            offset += 4;

            buffer[offset..offset + 4].copy_from_slice(&bounds.y().to_le_bytes());
            offset += 4;

            buffer[offset..offset + 4].copy_from_slice(&bounds.width().to_le_bytes());
            offset += 4;

            buffer[offset..offset + 4].copy_from_slice(&bounds.height().to_le_bytes());
            offset += 4;
        }

        buffer[offset..offset + 4].copy_from_slice(&(sources_size as u32).to_le_bytes());
        offset += 4;

        buffer[offset..offset + 4].copy_from_slice(&(placements_data_size as u32).to_le_bytes());
        offset += 4;

        // Write serialized sources
        buffer[offset..offset + sources_size].copy_from_slice(&serialized_sources);
        offset += sources_size;

        // Write placements data
        unsafe {
            std::ptr::copy_nonoverlapping(
                placements_data.as_ptr() as *const u8,
                buffer[offset..].as_mut_ptr(),
                placements_data_size,
            );
        }

        buffer
    }
}

struct BoundsData {
    bounds: BoundRect<f32>,
    result_bounds: BoundRect<f32>,
    area: f32,
    bin_node: PolygonNode,
}

#[derive(Debug, Clone)]
struct SourceItem {
    source: u16,
    children: Vec<SourceItem>,
}
